src/game_state.py

```diff
@@ -0,0 +1,588 @@
+        if self.cur_pitching_team.team_enum in team_pitch_event_map:
+            # Possible event, let's validate
+            event, start_season, end_season, req_blood = team_pitch_event_map[self.cur_pitching_team.team_enum]
+            # First let's check if a pre-pitch event for the pitcher should trigger
+            if team_pitch_event_map[self.cur_pitching_team.team_enum] in valid_pre_pitch_pitching_events:
+                # Let's figure out which pitch event it is and deal with it here
+
+                # Deal with charm strikeout chance
+                if event == PitchEventTeamBuff.CHARM and self.check_valid_season(start_season, end_season):
+                    if self.is_start_of_at_bat() and \
+                            self.check_blood_requirement(self.cur_pitching_team.starting_pitcher, req_blood):
+                        roll = random.random()
+                        if roll < CHARM_TRIGGER_PERCENTAGE:
+                            self.resolve_strikeout()
+                            return True
+                    return False
+                # TODO: Add additional pre pitch pitching events here as needed
+
+            # Second, let's check if a pre-pitch event for the batter should trigger
+            if team_pitch_event_map[self.cur_batting_team.team_enum] in valid_pre_pitch_batting_events:
+                # Let's figure out which batting event it is and deal with it here
+
+                # Deal with charm walk chance
+                if event == PitchEventTeamBuff.CHARM and self.check_valid_season(start_season, end_season):
+                    if self.is_start_of_at_bat() and \
+                            self.check_blood_requirement(self.cur_batting_team.cur_batter, req_blood):
+                        roll = random.random()
+                        if roll < CHARM_TRIGGER_PERCENTAGE:
+                            self.resolve_walk(1)
+                            return True
+                    return False
+
+                # Deal with zap chance
+                if event == PitchEventTeamBuff.ELECTRIC and self.check_valid_season(start_season, end_season):
+                    if self.is_start_of_at_bat() and \
+                            self.check_blood_requirement(self.cur_batting_team.cur_batter, req_blood):
+                        roll = random.random()
+                        if roll < ZAP_TRIGGER_PERCENTAGE:
+                            self.strikes -= 1
+                            return True
+                    return False
+                # TODO: Add additional pre pitch batting events here as needed
+
+        # Required checks failed, no event triggers
+        return False
+
+    def resolve_o_no(self) -> bool:
+        if self.cur_batting_team.team_enum in team_pitch_event_map:
+            event, start_season, end_season, req_blood = team_pitch_event_map[self.cur_batting_team.team_enum]
+            if event == PitchEventTeamBuff.O_NO:
+                if self.strikes == 2 and \
```

Comment:
this doesn't seem to be accounting for the number of balls in the count

Reply:
Good catch.  Fixed